pkg/ipmanager/ipam/controller.go

```diff
@@ -0,0 +1,236 @@
+	}, 60*time.Second)
+	config.PodInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: ipamController.enqueuePod,
+		UpdateFunc: func(old, new interface{}) {
+			ipamController.enqueuePod(new)
+		},
+		DeleteFunc: ipamController.handleRemovedPod,
+	}, 60*time.Second)
+
+	return ipamController
+}
+
+func (i *Controller) Run(stopCh <-chan struct{}) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer utilruntime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting IPAM Controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, i.servicesSynced, i.podsSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	klog.Info("Starting workers")
+	go wait.Until(i.runServiceWorker, time.Second, stopCh)
+	go wait.Until(i.runPodWorker, time.Second, stopCh)
+	wg.Wait()
+	<-stopCh
+	klog.Info("Shutting down workers")
+
+	return nil
+}
+
+func (i *Controller) runServiceWorker() {
+	for i.processNextService() {
+	}
+}
+
+func (i *Controller) runPodWorker() {
+	for i.processNextPod() {
+	}
+}
+
+const submarinerIpamGlobalIp = "submariner.io/globalIp"
+
+func (i *Controller) processNextService() bool {
+	obj, shutdown := i.serviceWorkqueue.Get()
+	if shutdown {
+		return false
+	}
+	err := func() error {
+		defer i.serviceWorkqueue.Done(obj)
+
+		key := obj.(string)
+		ns, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			i.serviceWorkqueue.Forget(obj)
+			return fmt.Errorf("error while splitting meta namespace key %s: %v", key, err)
+		}
+		if i.excludeNamespaces[ns] {
+			i.serviceWorkqueue.Forget(obj)
+			return nil
+		}
+
+		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of Service before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			service, err := i.kubeClientSet.CoreV1().Services(ns).Get(name, metav1.GetOptions{})
+
+			if err != nil {
+				i.serviceWorkqueue.Forget(obj)
+				return fmt.Errorf("error retrieving submariner-ipam-controller service object %s: %v", name, err)
+			}
+			annotations := service.GetAnnotations()
+			if annotations == nil {
+				annotations = map[string]string{}
+			}
+			if annotations[submarinerIpamGlobalIp] == "" {
+				annotations[submarinerIpamGlobalIp] = "169.254.1.1"
+				service.SetAnnotations(annotations)
+				klog.V(4).Infof("Updating service %s", service.Name)
+				_, updateErr := i.kubeClientSet.CoreV1().Services(ns).Update(service)
+				return updateErr
+			}
+			return nil
+		})
+		if retryErr != nil {
+			i.serviceWorkqueue.Forget(obj)
+			return retryErr
+		}
+
+		i.serviceWorkqueue.Forget(obj)
+		return nil
+	}()
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+func (i *Controller) processNextPod() bool {
```

Comment:
this function has an awful amount of repetition. Please re organize them into a function or method.